crates/bevy_asset/src/saver.rs

```diff
@@ -18,6 +19,7 @@ pub trait AssetSaver: Send + Sync + 'static {
         writer: &'a mut Writer,
         asset: SavedAsset<'a, Self::Asset>,
         settings: &'a Self::Settings,
+        ctx: &'a mut ProcessContext,
```

Comment:
I don't see anything inherently broken with this. But I'm a little curious about the use case here / I'm not sure I see the value yet. Adding the ability to consume arbitrary assets in the saver means that we're no longer "just" saving a loaded asset. Saver, in my mind, is a simple serializer of an already loaded asset. Introducing additional arbitrary loaded assets into the mix makes the dataflow less clear, and risks introducing behavior mismatches between processed and unprocessed versions of assets.

Can you walk me through the scenario that motivated this?

Reply:
Yeah, so I'm trying to use the asset v2 API for a workflow that compiles GLSL into SPIR-V. GLSL may contain `#include` statements which requires the asset system to load the included file as well. The included files must be registered as a dependency for the SPIR-V asset so that when you change the included files, all downstream shaders that included this file will be reloaded.

I'm running the `shaderc` compiler inside the asset saver, and I wouldn't know if the shader references any other files until that point. This is one of those cases where we can't "just" save a loaded asset - if we consider the compiler as a "serializer" from GLSL to SPIR-V, those additional loaded assets would then be files that are required only when the GLSL was saved.

Reply:
From my perspective loading and "compiling" should all happen in the AssetLoader so that the final resulting Shader asset is the "usable" shader (with all dependencies already resolved). The job of the Saver should just be to write the final `Shader` asset (ex: final fully resolved glsl, wgsl, or spirv) to a byte array.

This ensures that the Shader AssetLoader you are defining behaves the same (and is usable) in both processed and unprocessed mode. Enabling processing is just an optimization step to avoid resolving dependencies / loading directly from source files at runtime.

Reply:
The way I look at the asset system is that, each asset type represents a node in a Directed Acyclic Graph, and each "Processor" defines an edge in this DAG. As an end user, I should be able to request any node in the DAG, and the asset system will process the assets up till that point and give me exactly what I need.

What you're suggesting is that the `Saver` trait simply defines an optional "cacher" that does nothing more than simple serialization. The Loader that the "Saver" specifies would also do nothing more than simple deserialization. All the asset transformations should be done in a single `Loader`. I think that's fundamentally a less flexible model of the asset system. 

In my case for example, if my `GlslLoader` produces Spirv directly, I would not have an opportunity to load the GLSL source code directly in my application. On the other hand, a DAG formulation of the asset system allows the application to intervene at any level of the transformation pipeline. For example, I should be able to enable in-application shader editing, and the application will execute exactly the processors needed to turn this into Spirv which is what I want.



```
             --------------    GlslCompiler   ----------------
GlslLoader ->| GlslSource |  ---------------> | SpirvSource |  -- SpirvSaver
             --------------                   ----------------
```
versus
```
GlslLoader ---> | SpirvSource |  -- SpirvSaver
```

Reply:
> The way I look at the asset system is that, each asset type represents a node in a Directed Acyclic Graph, and each "Processor" defines an edge in this DAG. As an end user, I should be able to request any node in the DAG, and the asset system will process the assets up till that point and give me exactly what I need.

First: I agree with you on this. The `Process` trait is this unopinionated DAG edge and lets you request other asset data at any point during processing.

> What you're suggesting is that the Saver trait simply defines an optional "cacher" that does nothing more than simple serialization. The Loader that the "Saver" specifies would also do nothing more than simple deserialization. All the asset transformations should be done in a single Loader. I think that's fundamentally a less flexible model of the asset system.

That is essentially the idea. The `LoadAndSave` implementation of `Process` is an opinionated higher level abstraction that (from my current perspective) helps people build loaders and savers in a way that covers both processed and unprocessed asset loading scenarios. It encourages you to build an AssetLoader that can produce a "final" usable asset (Gltf -> Bevy Scene, Unresolved Bevy Shader Source -> Usable Shader).

If you are accessing additional asset information in a Saver to produce the final asset used by an app, that means you are building an engine feature that _requires_ Asset Preprocessing to work (aka something with a AssetAction::Process mode). We (or at least I) want to encourage people to build engine systems that do not require Asset Preprocessing to work (only things with AssetAction::Load), hence the restricted model.

Put another way, AssetLoader is "the part of the processing logic intended to be runnable at runtime without the AssetProcessor running in the background" and AssetSaver is "the part of the processing logic intended to be runnable only when the processor is running". AssetSaver is intentionally limited to encourage people to put the "smart" / "interesting" part that makes the asset "usable" into the AssetLoader. 

The shader system you are designing is one that fundamentally requires asset processing to be enabled (aka deploying and running AssetProcessor with the app) . This is ok, but its probably not the kind of thing we'd build into bevy by default, and its the kind of thing I'd want 3rd party developers to think twice about.

To fit the model above into the constraints of this system, I'd consider something like:

* Build multiple loaders (or multiple configurations for the same loader):
    *  A loader configuration that produces _only_ efficient fully loadable shader source (Ex: compiled Spirv). Trying to import a processed version of this shader source at runtime (via a Shader AssetLoader) would cause an error, as importing Spirv is not (currently) supported. 
    * A loader path that produces _both_ the efficient compiled shader source and the original source (ex: produces something like `Shader::spirv_with_glsl_source(spirv, glsl)`. Trying to import this processed shader at runtime (via a Shader AssetLoader) would work because we have the original source available.
    * A loader configuration that produces only uncompiled / unresolved shader source, for cases where a shader file is not intended to be used directly / will only be imported at runtime by other shaders.

Pretty sure all of these configurations fit into the LoadAndSave model.

> For example, I should be able to enable in-application shader editing, and the application will execute exactly the processors needed to turn this into Spirv which is what I want.

Note that this is a scenario I want to support. AssetProcessor is built to be deployable with Apps (and it already can be), and it in theory supports arbitrary AssetReader/AssetWriter backends. However it has currently only been tested with the filesystem backend (and things like the ProcessorTransactionLog still hard-code filesystem ops instead of putting them behind an interface). I strongly suspect there will be a few edge cases to work out before you could, say, produce in-memory assets to be processed by the AssetProcessor, which can then be loaded by the AssetServer (but this is something I'd like to support). And additionally, I'd like to explore the idea of running processors "inline" in the AssetServer without the AssetProcessor running (not sure yet if this is desirable / the right path forward). I consider this all "followup work" (see the Next Steps section of the PR).

So note that if the App you're building has a "user / app produces shaders on the file system, the AssetProcessor processes them, and the AssetServer loads the results" workflow in mind, it should work today. But if you have something like "in-memory-only processing" in mind, there will almost certainly be a few corner cases to work out still.

Reply:
Thanks for the in-depth response! I do still have some questions.

> The Process trait is this unopinionated DAG edge and lets you request other asset data at any point during processing.

However the `Process` trait feels very opinionated right now. For example, it gives you a `Writer`, which means that you're going to have to write something into the disk. This feels unnatural - what if I want to define a transformation between two asset types, but the second asset type is runtime only? I guess you could suggest that I should define two Asset loader, one that produces the first asset type directly, and another one that produces the second asset, and then another one that produces both. But as the asset processing pipeline gets longer, this requires exponentially more asset loader variants.

> We (or at least I) want to encourage people to build engine systems that do not require Asset Preprocessing to work (only things with AssetAction::Load), hence the restricted model.

I'm not sure this would be a good idea. Why build a feature, and then tell people "you probably shouldn't be using it all the time"? I understand that it's easier to get started with the "unprocessed" mode, but the presence of "processors" does not thwart this goal.

> Put another way, AssetLoader is "the part of the processing logic intended to be runnable at runtime without the AssetProcessor running in the background" and AssetSaver is "the part of the processing logic intended to be runnable only when the processor is running". AssetSaver is intentionally limited to encourage people to put the "smart" / "interesting" part that makes the asset "usable" into the AssetLoader.

This definition has a few problems. First, when I'm writing an `AssetLoader` I don't necessarily know if this is going to be run-time only, just like in the GLSL -> SPIRV example. I want to write the transformation once, and it's automatically flexible enough for both runtime and "process-time". Like we've explored, this definition requires that we have a very clear cut between "unprocessed" and "processed" mode, which causes confusion and just feels very unnatural.

Consider this alternative definition.
- AssetLoader: A simple deserializer that takes a `Reader` and produces in-memory asset objects. 
- AssetProcessor: A transformation that takes an in-memory asset object and produces another in-memory asset object.
- AssetSaver: A simple serializer that takes an in-memory asset object and write the data into a `Writer`.

By decoupling `AssetSaver` and `Processor` the dataflow becomes much more clearer. In the shader compiler example, we would only need to define four constructs: `GlslLoader`, `GlslToSpirvProcessor`, `SpirvSaver`, and `SpirvLoader`
- `GlslLoader`: Loads from a `.glsl` file, and write the string into `struct GlslSource { src: String }`
- `GlslToSpirvProcessor`: Invoke shaderc to compile GLSL into SPIRV. Input: `GlslSource`. Output: `struct SpirvSource { data: Vec<i32>}
- `SpirvSaver`: Write the raw data into a `Writer`.
- `SpirvLoader`: Read the raw data into `SpirvSource` directly.

We can then enable each component selectively  based on some compile-time feature flags:
- Unprocessed mode: Enable `GlslLoader`, `GlslToSpirvProcessor`.
- Processed mode: Enable `SpirvLoader`.
- Processed_dev mode: Enable `GlslLoader`, `GlslToSpirvProcessor`, `SpirvSaver`

Registered processors will always run, and they will try to recursively drive the DAG to completion as much as it can. If the user want to disable processing for certain assets, they just don't register them.

For me at least, this feels much more straightforward compared to "Build multiple loaders (or multiple configurations for the same loader)". We don't need to rewrite anything at all - just need to turn things on and off based on the feature flags enabled. And we certainly don't need to worry about writing too many asset loaders: no matter how long the asset pipeline might be, we can always elect to load an intermediate product and expect the processor to do the job.

Reply:
> And additionally, I'd like to explore the idea of running processors "inline" in the AssetServer without the AssetProcessor running (not sure yet if this is desirable / the right path forward). I consider this all "followup work" (see the Next Steps section of the PR).

That's certainly one way to run the processors at runtime, and imo should be the default way of running processors. When the user call "asset_server.load()", the asset server traverse the asset processing DAG backwards at runtime and figure out the dependency links. This would be equivalent to the "unprocessed" mode. If we're running on the "processed_dev" mode, we additionally run the `Saver` after the asset was processed. Our current mode of running processors (forward traversal) should be reserved for the special "cli mode", so that after it's executed, most assets will resolve immediately at runtime. This would be equivalent to the "processed" mode.

> AssetProcessor is built to be deployable with Apps (and it already can be), and it in theory supports arbitrary AssetReader/AssetWriter backends.

AssetProcessor shouldn't care about serializing the asset. It's a transformation that turns one asset type into another. Writing in to a `Writer` requires serialization which isn't the job of a processor.

Reply:
> However the Process trait feels very opinionated right now.

I think the biggest disconnect here is phrasing. Think of the Process trait (and the "processing" term) not as the generic "runtime asset transformation trait", but instead as "the path from source asset bytes to destination asset bytes". From the AssetServer perspective, it receives a request for a AssetPath key, it feeds that into an AssetReader, which returns bytes, which are fed into a configured AssetLoader. Preprocessing's job is to just change what bytes are returned and configure what loader to use.

Process doesn't care about in memory `Asset` state. It cares about producing bytes that the asset server can load for a given AssetPath.

From this perspective, I see nothing less opinionated than Process (for the scenario of preprocessing). I think it is important to have a "input and output byte only" view of the world for preprocessing, as we are doing things like "hashing source files to determine if they have changed" and "deleting output files if they are no longer necessary".

All of the "transformation chaining" logic that you are describing (ex: mappings from GlslSource to SpirVSource) are things that exist "between" the source and the destination bytes.

Think of the `Process` trait as `Preprocess`, if you would like to reserve the "process" term for "in memory processing" instead of "asset source processing" (but to avoid confusion i'd prefer it if we used the existing terminology for now and called what you're discussing "asset transforms").

> Consider this alternative definition

> * AssetLoader: A simple deserializer that takes a Reader and produces in-memory asset objects.

Perfect match ... this is already what AssetLoader is.

> AssetProcessor: A transformation that takes an in-memory asset object and produces another in-memory asset object.

This is roughly what I called "asset transforms" in the **More high level Processor trait implementations** Next Steps section of the PR. Seems like something that can fit into the existing system without reframing loaders or savers or preprocessors. Running transforms for "non-pre-processed assets" is totally fair game / is also a generally compatible add-on to the system as it exists today.

I see this as being roughly AssetLoader-like in api (aka you can still kick off additional asset loads + it adds to the tracked dependencies of the asset), but instead of having a byte Reader it has the input asset type.

> AssetSaver: A simple serializer that takes an in-memory asset object and write the data into a Writer.
> By decoupling AssetSaver and Processor the dataflow becomes much more clearer.

The AssetSaver you've defined above is the AssetSaver that exists today. AssetSaver and Process (again think of this as PreProcess in your terms) as built today _are not logically coupled_. The only potential coupling is if we add ProcessContext to the interface, which I'll note is what you are proposing in this PR and that I pushed back against :)

> I'm not sure this would be a good idea. Why build a feature, and then tell people "you probably shouldn't be using it all the time"? I understand that it's easier to get started with the "unprocessed" mode, but the presence of "processors" does not thwart this goal.

Bevy has a _ton_ of optional features. Not everyone wants to (or should be) using Temporal Anti Aliasing, so we make it optional. Some game developers don't want to worry about meta files, managing input and output asset folders, waiting for assets to compile, etc. They just want to load their PNG sprites directly and call it a day.

And possibly more importantly, you will note that "processed" and "unprocessed" mode is essentially identical (they only point to different asset folders). The only mode that adds additional "stuff" is "processed-dev" (aka, the mode that enables the AssetProcessor, which does "preprocessing"). _Most_ games do not need to be deployed with the AssetProcessor code (ex: deployed games generally wont need the writeahead transaction log). And most games also don't need to be deployed with things like mesh optimizers, Basis Universal encoders, etc. Having distinct on/off preprocessed/non-preprocessed modes (and encouraging users to think about it this way) helps frame the situation properly.

> Why make a distinction between Asset Preprocessing and Asset Processing? Assets are always processed one way or another. The only difference is that "preprocessing" caches the results to the disk, so that you don't have to process them again.

Because "preprocessing" is the very specific action of scanning through an entire source folder, processing all assets to completion, and writing the outputs to a destination folder. It involves doing extra work such as checking source files to see if they changed and comparing them to stored output files. In a "preprocessing" configuration, when you deploy your game the source assets aren't even available anymore.

There is a reason AssetProcessor (rename this AssetPreprocessor in your head if it helps) looks different than AssetServer. It tracks different metadata, does different work, and has different performance characteristics + constraints. The process of "just" loading (and transforming) assets at runtime _looks different_ and has different design requirements than "asset preprocessing". Most deployed games don't need (or want) a "write ahead transaction log".

> Registered processors will always run

I think this needs to be configurable per-asset, with engine defined defaults, user defined default overrides, and per-asset overrides in meta files (aka the meta files and global per-asset-type defaults we have today).

Reply:
In general my take is:

* The current design is already (roughly) what it should be as a foundation
* We _both_ want to add something like a `AssetTransform(A) -> B`, which can be configured for assets like:
    * `AssetAction::Load(GlslLoader->GlslCompiler)` (run the compiler at runtime after loading)
    * `AssetAction::Process(GlslLoader->GlslCompiler->SpirvSaver)`(run the compiler and save the results at "preprocess time")
* The `AssetTransform` as described above fits into the existing system / could be accomplished as a "next step" after the BA V2 pr lands. There is already a Next Step item that touches on this, but not as clearly (or as fleshed out) as it is expressed above.
* We also both want true "one-to-many" processed assets (as in labeled sub assets each have their own output files), with each labeled asset capable of having its own configured transform/save chain. This is already a "next step" defined in the PR.
* Terminology is flexible. We can do AssetTransform->AssetProcessor, AssetProcessor->AssetPreprocessor, etc. I think the current terms are fine, but happy to discuss further.